File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <stack>

using namespace std;

stack<char>st;

int solve(string s){
    
    for(int i=0;i<s.size();i++){
        char c = s[i];
        
        if(c==')' || c==']' || c=='}'){
            if(st.empty()) return 0;
        }
        if(c=='(' || c=='[' || c=='{'){
            st.push(c);
        }else if(c==')'){
           if(st.top()=='('){
               st.pop();
           } else{
               return 0;
           }
        }else if(c==']'){
            if(st.top()=='['){
               st.pop();
           } else{
               return 0;
           }
        }else if(c=='}'){
            if(st.top()=='{'){
               st.pop();
           } else{
               return 0;
           }
        }
    }
    
    if(st.empty()){
        return 1;
    }else{
        while(!st.empty()) st.pop();
        return 0;
    }
}

int solution(string s) {
    int answer = 0;
    
    answer += solve(s);
    
    for(int i=0;i<s.size()-1;i++){
        s = s.substr(1) + s[0];
        answer += solve(s);
    }
    
    return answer;
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.Stack;

class Solution {
    
    public int solution(String s) {
        int answer = 0;
        
        int n = s.length();
        
        s+=s;
        
        for(int i=0;i<n;i++){
            Stack<Character> stack = new Stack<>();
            int check =1;
            for(int j=i;j<i+n;j++){
                char c = s.charAt(j);
                if(c=='(' || c=='[' || c=='{'){
                    stack.push(c);
                }else{
                    if(stack.empty()){
                        check=0;
                        break;
                    }else{
                        char e = stack.peek();
                        if((e=='(' && c==')') || (e=='[' && c==']') || 
                          (e=='{' && c=='}')){
                            stack.pop();
                                continue;
                        }else{
                            check=0;
                            break;
                        }
                    }
                }
            }
            if(stack.empty() && check==1) answer++;
        }
    
        
        return answer;
    }
}